src/hooks/room/useRoomCallStatus.ts

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Room, RoomStateEvent } from "matrix-js-sdk/src/matrix";
+import { useCallback, useEffect, useMemo, useState } from "react";
+
+import { useFeatureEnabled } from "../useSettings";
+import SdkConfig from "../../SdkConfig";
+import { useEventEmitterState, useTypedEventEmitterState } from "../useEventEmitter";
+import LegacyCallHandler, { LegacyCallHandlerEvent } from "../../LegacyCallHandler";
+import { useWidgets } from "../../components/views/right_panel/RoomSummaryCard";
+import { WidgetType } from "../../widgets/WidgetType";
+import { useCall } from "../useCall";
+import { _t } from "../../languageHandler";
+import { useRoomMembers } from "../useRoomMembers";
+import { ElementCall } from "../../models/Call";
+
+type CallType = "element_call" | "jitsi_or_element_call" | "legacy_or_jitsi";
+
+/**
+ * Reports the call capabilities for the current room
+ * @param room the room to track
+ * @returns the call status for a room
+ */
+export const useRoomCallStatus = (
+    room: Room,
+): {
+    voiceCallDisabledReason: string | null;
+    voiceCallType: CallType;
+    videoCallDisabledReason: string | null;
+    videoCallType: CallType;
+} => {
+    const [voiceCallDisabledReason, setVoiceCallDisabledReason] = useState<string | null>(null);
+    const [videoCallDisabledReason, setVideoCallDisabledReason] = useState<string | null>(null);
+    const [voiceCallType, setVoiceCallType] = useState<CallType>("jitsi_or_element_call");
+    const [videoCallType, setVideoCallType] = useState<CallType>("jitsi_or_element_call");
+
+    const groupCallsEnabled = useFeatureEnabled("feature_group_calls");
+    const useElementCallExclusively = useMemo(() => {
+        return SdkConfig.get("element_call").use_exclusively;
+    }, []);
+
+    const hasLegacyCall = useEventEmitterState(
+        LegacyCallHandler.instance,
+        LegacyCallHandlerEvent.CallsChanged,
+        useCallback(() => LegacyCallHandler.instance.getCallForRoom(room.roomId) !== null, [room]),
+    );
+
+    const widgets = useWidgets(room);
+    const hasJitsiWidget = useMemo(() => widgets.some((widget) => WidgetType.JITSI.matches(widget.type)), [widgets]);
+
+    const hasGroupCall = useCall(room.roomId) !== null;
+
+    const functionalMembers = useRoomMembers(room);
```

Comment:
What does `functional` mean here? Aren't his just all joined room members?

Reply:
The logic is ported over from https://github.com/matrix-org/matrix-react-sdk/blob/bba980cd8a55e4c9bdf18eaec3ab29174b6a73da/src/components/views/rooms/LegacyRoomHeader.tsx#L267

It seems that a functional member is a member that is not a bot according to this JSDoc comment,

```
all room members that are non-functional (bots etc.).
```

Reply:
`useRoomMembers` does something different here. The legacy code is also annoying, because it assigned the result of `getJoinedNonFunctionalMembers` to `functionalMembers`.

→ It should be `nonFunctionalMembers` (in the old code, out of scope here but also here in the new code)